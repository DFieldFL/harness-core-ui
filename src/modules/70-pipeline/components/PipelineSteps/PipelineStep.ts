import { omit } from 'lodash-es'

import { Step, StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { isCustomGeneratedString } from '@pipeline/components/PipelineStudio/ExecutionGraph/ExecutionGraphUtil'
export type { StepProps } from '@pipeline/components/AbstractSteps/Step'

export abstract class PipelineStep<T extends { name?: string; identifier?: string }> extends Step<T> {
  getDefaultValues(initialValues: T, viewType: StepViewType): T {
    if (initialValues.identifier && isCustomGeneratedString(initialValues.identifier)) {
      return { ...this.defaultValues, ...omit(initialValues, ['name', 'identifier']) }
    }
    if (viewType === StepViewType.InputSet || viewType === StepViewType.DeploymentForm) {
      return {
        ...initialValues
      }
    }
    return {
      ...this.defaultValues,
      ...initialValues
    }
  }
  processFormData(values: T): T {
    return values
  }
}
