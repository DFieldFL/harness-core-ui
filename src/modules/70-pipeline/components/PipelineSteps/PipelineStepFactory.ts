import { AbstractStepFactory } from '@pipeline/exports'

import { StepGroupStep } from './Steps/StepGroupStep/StepGroupStep'
import { CustomVariables } from './Steps/CustomVariables/CustomVariables'
import { BarrierStep } from './Steps/Barrier/Barrier'
import { HarnessApproval } from './Steps/Approval/HarnessApproval'
import { JiraApproval } from './Steps/JiraApproval/JiraApproval'
import { JiraCreate } from './Steps/JiraCreate/JiraCreate'
import { JiraUpdate } from './Steps/JiraUpdate/JiraUpdate'

class PipelineStepFactory extends AbstractStepFactory {
  protected type = 'pipeline-factory'
}

const factory = new PipelineStepFactory()

// common
factory.registerStep(new BarrierStep())
factory.registerStep(new StepGroupStep())
factory.registerStep(new CustomVariables())
factory.registerStep(new HarnessApproval())
factory.registerStep(new JiraApproval())
factory.registerStep(new JiraCreate())
factory.registerStep(new JiraUpdate())

// build steps
export default factory
