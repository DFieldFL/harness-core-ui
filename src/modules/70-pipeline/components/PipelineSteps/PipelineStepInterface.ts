export enum StepType {
  HTTP = 'Http',
  SHELLSCRIPT = 'ShellScript',
  APPROVAL = 'Approval',
  K8sRollingRollback = 'K8sRollingRollback',
  StepGroup = 'StepGroup',
  KubernetesInfraSpec = 'KubernetesInfraSpec',
  K8sRollingDeploy = 'K8sRollingDeploy',
  CustomVariable = 'Custom_Variable',
  Redis = 'service',
  Plugin = 'plugin',
  Run = 'run',
  SaveCache = 'Save_Cache',
  RestoreCache = 'Restore_Cache',
  GCR = 'GCR'
}
