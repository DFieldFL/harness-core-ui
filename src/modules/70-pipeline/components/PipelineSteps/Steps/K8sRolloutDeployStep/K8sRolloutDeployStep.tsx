import React from 'react'
import {
  IconName,
  Formik,
  FormInput,
  Button,
  getMultiTypeFromValue,
  MultiTypeInputType,
  Accordion,
  Text
} from '@wings-software/uicore'
import * as Yup from 'yup'

import { FormikProps, yupToFormErrors } from 'formik'
import { isEmpty } from 'lodash-es'
import { StepViewType, StepProps } from '@pipeline/exports'
import type { K8sRollingStepInfo, StepElement } from 'services/cd-ng'
import { FormMultiTypeCheckboxField } from '@common/components'
import { ConfigureOptions } from '@common/components/ConfigureOptions/ConfigureOptions'
import { setFormikRef } from '@pipeline/components/AbstractSteps/Step'
import type { StepFormikFowardRef } from '@pipeline/components/AbstractSteps/Step'
import { useStrings, UseStringsReturn } from 'framework/exports'
import {
  DurationInputFieldForInputSet,
  FormMultiTypeDurationField,
  getDurationValidationSchema
} from '@common/components/MultiTypeDuration/MultiTypeDuration'
import type { VariableMergeServiceResponse } from 'services/pipeline-ng'
import { CopyText } from '@common/components/CopyText/CopyText'
import { toVariableStr } from '@common/utils/StringUtils'
import { StepType } from '../../PipelineStepInterface'
import { PipelineStep } from '../../PipelineStep'
import stepCss from '../Steps.module.scss'
import variableCss from '@pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.module.scss'

export interface K8RolloutDeployData extends StepElement {
  spec: K8sRollingStepInfo
}

interface K8RolloutDeployProps {
  initialValues: K8RolloutDeployData
  onUpdate?: (data: K8RolloutDeployData) => void
  stepViewType?: StepViewType
  inputSetData?: {
    template?: K8RolloutDeployData
    path?: string
    readonly?: boolean
  }
}

function K8RolloutDeployWidget(
  props: K8RolloutDeployProps,
  formikRef: StepFormikFowardRef<K8RolloutDeployData>
): React.ReactElement {
  const { initialValues, onUpdate } = props
  const { getString } = useStrings()
  return (
    <>
      <Formik<K8RolloutDeployData>
        onSubmit={(values: K8RolloutDeployData) => {
          onUpdate?.(values)
        }}
        initialValues={initialValues}
        validationSchema={Yup.object().shape({
          name: Yup.string().required(getString('pipelineSteps.stepNameRequired')),
          spec: Yup.object().shape({
            timeout: getDurationValidationSchema({ minimum: '10s' }).required(
              getString('validation.timeout10SecMinimum')
            )
          })
        })}
      >
        {(formik: FormikProps<K8RolloutDeployData>) => {
          setFormikRef(formikRef, formik)
          const { values, submitForm, setFieldValue } = formik
          return (
            <>
              <Accordion activeId="details" collapseProps={{ transitionDuration: 0 }}>
                <Accordion.Panel
                  id="details"
                  summary={getString('pipelineSteps.k8sRolloutDeploy')}
                  details={
                    <>
                      <div className={stepCss.formGroup}>
                        <FormInput.InputWithIdentifier inputLabel={getString('name')} />
                      </div>
                      <div className={stepCss.formGroup}>
                        <FormMultiTypeDurationField
                          name="spec.timeout"
                          label={getString('pipelineSteps.timeoutLabel')}
                          multiTypeDurationProps={{ enableConfigureOptions: false }}
                        />
                        {getMultiTypeFromValue(values.spec.timeout) === MultiTypeInputType.RUNTIME && (
                          <ConfigureOptions
                            value={values.spec.timeout as string}
                            type="String"
                            variableName="step.spec.timeout"
                            showRequiredField={false}
                            showDefaultField={false}
                            showAdvanced={true}
                            onChange={value => {
                              setFieldValue('spec.timeout', value)
                            }}
                          />
                        )}
                      </div>
                      <div className={stepCss.formGroup}>
                        <FormMultiTypeCheckboxField
                          name="spec.skipDryRun"
                          label={getString('pipelineSteps.skipDryRun')}
                        />
                      </div>
                    </>
                  }
                />
              </Accordion>
              <div className={stepCss.actionsPanel}>
                <Button intent="primary" text={getString('submit')} onClick={submitForm} />
              </div>
            </>
          )
        }}
      </Formik>
    </>
  )
}

const K8RolloutDeployInputStep: React.FC<K8RolloutDeployProps> = ({ inputSetData }) => {
  const { getString } = useStrings()
  return (
    <>
      {getMultiTypeFromValue(inputSetData?.template?.spec?.timeout) === MultiTypeInputType.RUNTIME && (
        <DurationInputFieldForInputSet
          label={getString('pipelineSteps.timeoutLabel')}
          name={`${isEmpty(inputSetData?.path) ? '' : `${inputSetData?.path}.`}spec.timeout`}
          disabled={inputSetData?.readonly}
        />
      )}
      {getMultiTypeFromValue(inputSetData?.template?.spec?.skipDryRun) === MultiTypeInputType.RUNTIME && (
        <FormInput.CheckBox
          name={`${isEmpty(inputSetData?.path) ? '' : `${inputSetData?.path}.`}spec.skipDryRun`}
          className={stepCss.checkbox}
          label={getString('pipelineSteps.skipDryRun')}
          disabled={inputSetData?.readonly}
        />
      )}
    </>
  )
}
export interface K8RolloutDeployVariableStepProps {
  initialValues: K8RolloutDeployData
  stageIdentifier: string
  onUpdate?(data: K8RolloutDeployData): void
  metadataMap: Required<VariableMergeServiceResponse>['metadataMap']
  variablesData: K8RolloutDeployData
}

const K8RolloutDeployVariableStep: React.FC<K8RolloutDeployVariableStepProps> = ({
  variablesData,
  metadataMap,
  initialValues
}) => {
  return (
    <div className={variableCss.variableListTable}>
      <>
        <CopyText
          textToCopy={toVariableStr(
            metadataMap[(variablesData.spec.skipDryRun as unknown) as string].yamlProperties?.fqn || ''
          )}
        >
          &lt;+{metadataMap[(variablesData.spec.skipDryRun as unknown) as string].yamlProperties?.localName || ''}&gt;
        </CopyText>
        <Text>Boolean</Text>
        <Text>{initialValues.spec.skipDryRun === true ? 'true' : 'false'}</Text>
      </>
      <>
        <CopyText
          textToCopy={toVariableStr(
            metadataMap[(variablesData.spec.timeout as unknown) as string].yamlProperties?.fqn || ''
          )}
        >
          &lt;+{metadataMap[(variablesData.spec.timeout as unknown) as string].yamlProperties?.localName || ''}&gt;
        </CopyText>
        <Text>Boolean</Text>
        <Text>{initialValues.spec.timeout}</Text>
      </>
    </div>
  )
}

const K8sRolloutDeployRef = React.forwardRef(K8RolloutDeployWidget)
export class K8RolloutDeployStep extends PipelineStep<K8RolloutDeployData> {
  constructor() {
    super()
    this._hasStepVariables = true
  }
  renderStep(props: StepProps<K8RolloutDeployData>): JSX.Element {
    const { initialValues, onUpdate, stepViewType, inputSetData, formikRef, customStepProps } = props

    if (stepViewType === StepViewType.InputSet || stepViewType === StepViewType.DeploymentForm) {
      return (
        <K8RolloutDeployInputStep
          initialValues={initialValues}
          onUpdate={onUpdate}
          stepViewType={stepViewType}
          inputSetData={inputSetData}
        />
      )
    } else if (stepViewType === StepViewType.InputVariable) {
      return (
        <K8RolloutDeployVariableStep
          {...(customStepProps as K8RolloutDeployVariableStepProps)}
          initialValues={initialValues}
          onUpdate={onUpdate}
        />
      )
    }

    return (
      <K8sRolloutDeployRef
        initialValues={initialValues}
        onUpdate={onUpdate}
        stepViewType={stepViewType}
        ref={formikRef}
      />
    )
  }
  validateInputSet(
    data: K8RolloutDeployData,
    template: K8RolloutDeployData,
    getString?: UseStringsReturn['getString']
  ): object {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const errors = { spec: {} } as any
    if (getMultiTypeFromValue(template?.spec?.timeout) === MultiTypeInputType.RUNTIME) {
      const timeout = Yup.object().shape({
        timeout: getDurationValidationSchema({ minimum: '10s' }).required(getString?.('validation.timeout10SecMinimum'))
      })

      try {
        timeout.validateSync(data.spec)
      } catch (e) {
        /* istanbul ignore else */
        if (e instanceof Yup.ValidationError) {
          const err = yupToFormErrors(e)

          Object.assign(errors.spec, err)
        }
      }
    }
    /* istanbul ignore else */
    if (isEmpty(errors.spec)) {
      delete errors.spec
    }

    return errors
  }

  protected type = StepType.K8sRollingDeploy
  protected stepName = 'K8s Rollout Deploy'
  protected stepIcon: IconName = 'service-kubernetes'

  protected defaultValues: K8RolloutDeployData = {
    identifier: '',
    spec: {
      skipDryRun: false,
      timeout: '10m'
    }
  }
}
