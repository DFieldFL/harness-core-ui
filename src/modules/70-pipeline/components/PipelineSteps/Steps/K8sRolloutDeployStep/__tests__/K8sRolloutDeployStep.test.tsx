import React from 'react'
import { render } from '@testing-library/react'
import { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { StepType } from '@pipeline/components/PipelineSteps/PipelineStepInterface'
import { factory, TestStepWidget } from '../../__tests__/StepTestUtil'
import { K8RolloutDeployStep } from '../K8sRolloutDeployStep'

describe('Test K8sRolloutDeployStep', () => {
  beforeEach(() => {
    factory.registerStep(new K8RolloutDeployStep())
  })
  test('should render edit view as new step', () => {
    const { container } = render(
      <TestStepWidget initialValues={{}} type={StepType.K8sRollingDeploy} stepViewType={StepViewType.Edit} />
    )
    expect(container).toMatchSnapshot()
  })
  test('should render edit view as edit step', () => {
    const { container } = render(
      <TestStepWidget
        initialValues={{
          type: 'K8sRollingDeploy',
          name: 'Test A',
          identifier: 'Test_A',
          spec: { skipDryRun: '${input}', timeout: '10m' }
        }}
        type={StepType.K8sRollingDeploy}
        stepViewType={StepViewType.Edit}
      />
    )
    expect(container).toMatchSnapshot()
  })
  test('should render edit view', () => {
    const { container } = render(
      <TestStepWidget
        initialValues={{ identifier: 'Test_A', type: 'K8sRollingDeploy', spec: { skipDryRun: false } }}
        template={{ identifier: 'Test_A', type: 'K8sRollingDeploy', spec: { skipDryRun: '${input}' } }}
        allValues={{
          type: 'K8sRollingDeploy',
          name: 'Test A',
          identifier: 'Test_A',
          spec: { skipDryRun: '${input}', timeout: '10m' }
        }}
        type={StepType.K8sRollingDeploy}
        stepViewType={StepViewType.InputSet}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
