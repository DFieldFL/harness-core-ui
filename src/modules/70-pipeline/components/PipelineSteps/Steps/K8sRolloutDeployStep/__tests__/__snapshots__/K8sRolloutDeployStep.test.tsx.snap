// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test K8sRolloutDeployStep should render edit view 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <div
      class="bp3-form-group"
    >
      <div
        class="bp3-form-content"
      >
        <label
          class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox Checkbox--checkbox checkbox StyledProps--main StyledProps--inline"
        >
          <input
            name="spec.skipDryRun"
            type="checkbox"
          />
          <span
            class="bp3-control-indicator"
          />
          Skip Dry Run
        </label>
      </div>
    </div>
    <p
      class="StyledProps--font StyledProps--main"
    >
      Errors
    </p>
    <pre>
      {}
    </pre>
    <button
      class="bp3-button StyledProps--font Button--button StyledProps--main StyledProps--intent-primary"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Submit
      </span>
    </button>
  </div>
</div>
`;

exports[`Test K8sRolloutDeployStep should render edit view as edit step 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <div
      class="Accordion--accordion"
    >
      <div
        class="Accordion--panel"
        data-open="true"
        data-testid="details-panel"
      >
        <div
          class="Accordion--summary"
          data-testid="details-summary"
        >
          <div
            class="Accordion--label"
          >
            K8s Rollout Deploy
          </div>
          <div
            class="Accordion--chevron"
          />
        </div>
        <div
          class="bp3-collapse"
          style="height: auto; overflow-y: visible; transition: none;"
        >
          <div
            aria-hidden="false"
            class="bp3-collapse-body"
            style="transform: translateY(0); transition: none;"
          >
            <div
              class="Accordion--details"
              data-testid="details-details"
            >
              <div
                class="formGroup"
              >
                <div
                  class="InputWithIdentifier--txtNameContainer"
                >
                  <div
                    class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main InputWithIdentifier--txtIdContainer"
                  >
                    <p
                      class="StyledProps--font StyledProps--main"
                    >
                      ID
                      :
                    </p>
                    <span
                      class="bp3-popover-wrapper"
                    >
                      <span
                        class="bp3-popover-target"
                      >
                        <div
                          class="bp3-editable-text InputWithIdentifier--idValue"
                          tabindex="0"
                        >
                          <span
                            class="bp3-editable-text-content"
                            style="height: 1px; line-height: 1px; min-width: 0;"
                          >
                            Test_A
                          </span>
                        </div>
                      </span>
                    </span>
                    <span
                      class="bp3-icon bp3-icon-edit StyledProps--main"
                      icon="edit"
                      style="vertical-align: middle;"
                    >
                      <svg
                        data-icon="edit"
                        height="12"
                        viewBox="0 0 16 16"
                        width="12"
                      >
                        <desc>
                          edit
                        </desc>
                        <path
                          d="M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </div>
                  <div
                    class="bp3-form-group"
                  >
                    <label
                      class="bp3-label"
                      for="name"
                    >
                      Name
                       
                      <span
                        class="bp3-text-muted"
                      />
                    </label>
                    <div
                      class="bp3-form-content"
                    >
                      <div
                        class="bp3-input-group"
                      >
                        <input
                          autocomplete="off"
                          class="bp3-input"
                          name="name"
                          type="text"
                          value="Test A"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="formGroup"
              >
                <div
                  class="bp3-form-group"
                >
                  <label
                    class="bp3-label"
                    for="spec.timeout"
                  >
                    Timeout
                     
                    <span
                      class="bp3-text-muted"
                    />
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main MultiTypeInput--main"
                      style="flex-grow: 1;"
                    >
                      <div
                        class="bp3-input-group bp3-fill"
                      >
                        <input
                          class="bp3-input"
                          name="spec.timeout"
                          placeholder="Enter w/d/h/m/s/ms"
                          type="text"
                          value="10m"
                        />
                      </div>
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="MultiTypeInput--btn MultiTypeInput--FIXED"
                          >
                            <span
                              class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-white"
                              data-icon="fixed-input"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 9 9"
                                width="12"
                              >
                                <path
                                  d="M5.61 7.55c.62-.7.3-1.42.18-2.35a.32.32 0 01.1-.29l1.43-1.42a.76.76 0 01.28-.15c.39-.1 1 0 1.28-.29a.34.34 0 000-.51L6.46.12a.33.33 0 00-.52 0c-.28.29-.2.9-.28 1.3a.53.53 0 01-.13.23L4.09 3.12a.29.29 0 01-.28.09c-.89-.11-1.6-.49-2.27.11-.29.26-.29.43 0 .7l1.31 1.34.15.16-.15.15-2.7 2.69a.36.36 0 000 .54.36.36 0 00.54-.05l2.63-2.68.17-.17.17.16c.45.46.9.91 1.36 1.36.21.21.42.22.59.03z"
                                  fill="#009dff"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="formGroup"
              >
                <div
                  class="bp3-form-group"
                >
                  <label
                    class="bp3-label"
                    for="spec.skipDryRun"
                  >
                    Skip Dry Run
                     
                    <span
                      class="bp3-text-muted"
                    />
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                    >
                      <div
                        class="TextInput--main MultiTypeInput--input"
                      >
                        <div
                          class="bp3-input-group bp3-disabled"
                        >
                          <input
                            class="bp3-input"
                            disabled=""
                            name="spec.skipDryRun"
                            placeholder="<+input>"
                            type="text"
                            value="<+input>"
                          />
                        </div>
                      </div>
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                          >
                            <span
                              class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-white"
                              data-icon="runtime-input"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 10 7.97"
                                width="12"
                              >
                                <path
                                  d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                                  fill="#9d7cff"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="actionsPanel"
    >
      <button
        class="bp3-button StyledProps--font Button--button StyledProps--main StyledProps--intent-primary"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Submit
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Test K8sRolloutDeployStep should render edit view as new step 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <div
      class="Accordion--accordion"
    >
      <div
        class="Accordion--panel"
        data-open="true"
        data-testid="details-panel"
      >
        <div
          class="Accordion--summary"
          data-testid="details-summary"
        >
          <div
            class="Accordion--label"
          >
            K8s Rollout Deploy
          </div>
          <div
            class="Accordion--chevron"
          />
        </div>
        <div
          class="bp3-collapse"
          style="height: auto; overflow-y: visible; transition: none;"
        >
          <div
            aria-hidden="false"
            class="bp3-collapse-body"
            style="transform: translateY(0); transition: none;"
          >
            <div
              class="Accordion--details"
              data-testid="details-details"
            >
              <div
                class="formGroup"
              >
                <div
                  class="InputWithIdentifier--txtNameContainer"
                >
                  <div
                    class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main InputWithIdentifier--txtIdContainer"
                  >
                    <p
                      class="StyledProps--font StyledProps--main"
                    >
                      ID
                      :
                    </p>
                    <span
                      class="bp3-popover-wrapper"
                    >
                      <span
                        class="bp3-popover-target"
                      >
                        <div
                          class="bp3-editable-text bp3-editable-text-placeholder InputWithIdentifier--idValue"
                          tabindex="0"
                        >
                          <span
                            class="bp3-editable-text-content"
                            style="height: 1px; line-height: 1px; min-width: 0;"
                          />
                        </div>
                      </span>
                    </span>
                    <span
                      class="bp3-icon bp3-icon-edit StyledProps--main"
                      icon="edit"
                      style="vertical-align: middle;"
                    >
                      <svg
                        data-icon="edit"
                        height="12"
                        viewBox="0 0 16 16"
                        width="12"
                      >
                        <desc>
                          edit
                        </desc>
                        <path
                          d="M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </div>
                  <div
                    class="bp3-form-group"
                  >
                    <label
                      class="bp3-label"
                      for="name"
                    >
                      Name
                       
                      <span
                        class="bp3-text-muted"
                      />
                    </label>
                    <div
                      class="bp3-form-content"
                    >
                      <div
                        class="bp3-input-group"
                      >
                        <input
                          autocomplete="off"
                          class="bp3-input"
                          name="name"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="formGroup"
              >
                <div
                  class="bp3-form-group"
                >
                  <label
                    class="bp3-label"
                    for="spec.timeout"
                  >
                    Timeout
                     
                    <span
                      class="bp3-text-muted"
                    />
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main MultiTypeInput--main"
                      style="flex-grow: 1;"
                    >
                      <div
                        class="bp3-input-group bp3-fill"
                      >
                        <input
                          class="bp3-input"
                          name="spec.timeout"
                          placeholder="Enter w/d/h/m/s/ms"
                          type="text"
                          value="10m"
                        />
                      </div>
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="MultiTypeInput--btn MultiTypeInput--FIXED"
                          >
                            <span
                              class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-white"
                              data-icon="fixed-input"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 9 9"
                                width="12"
                              >
                                <path
                                  d="M5.61 7.55c.62-.7.3-1.42.18-2.35a.32.32 0 01.1-.29l1.43-1.42a.76.76 0 01.28-.15c.39-.1 1 0 1.28-.29a.34.34 0 000-.51L6.46.12a.33.33 0 00-.52 0c-.28.29-.2.9-.28 1.3a.53.53 0 01-.13.23L4.09 3.12a.29.29 0 01-.28.09c-.89-.11-1.6-.49-2.27.11-.29.26-.29.43 0 .7l1.31 1.34.15.16-.15.15-2.7 2.69a.36.36 0 000 .54.36.36 0 00.54-.05l2.63-2.68.17-.17.17.16c.45.46.9.91 1.36 1.36.21.21.42.22.59.03z"
                                  fill="#009dff"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="formGroup"
              >
                <div
                  class="bp3-form-group"
                >
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main MultiTypeInput--main"
                    >
                      <label
                        class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox input StyledProps--main"
                      >
                        <input
                          name="spec.skipDryRun"
                          type="checkbox"
                        />
                        <span
                          class="bp3-control-indicator"
                        />
                        Skip Dry Run
                      </label>
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="MultiTypeInput--btn MultiTypeInput--FIXED"
                          >
                            <span
                              class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-white"
                              data-icon="fixed-input"
                            >
                              <svg
                                height="12"
                                viewBox="0 0 9 9"
                                width="12"
                              >
                                <path
                                  d="M5.61 7.55c.62-.7.3-1.42.18-2.35a.32.32 0 01.1-.29l1.43-1.42a.76.76 0 01.28-.15c.39-.1 1 0 1.28-.29a.34.34 0 000-.51L6.46.12a.33.33 0 00-.52 0c-.28.29-.2.9-.28 1.3a.53.53 0 01-.13.23L4.09 3.12a.29.29 0 01-.28.09c-.89-.11-1.6-.49-2.27.11-.29.26-.29.43 0 .7l1.31 1.34.15.16-.15.15-2.7 2.69a.36.36 0 000 .54.36.36 0 00.54-.05l2.63-2.68.17-.17.17.16c.45.46.9.91 1.36 1.36.21.21.42.22.59.03z"
                                  fill="#009dff"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="actionsPanel"
    >
      <button
        class="bp3-button StyledProps--font Button--button StyledProps--main StyledProps--intent-primary"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Submit
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Test K8sRolloutDeployStep should render variable view 1`] = `
<div>
  <div
    class="variablesListTable"
  >
    <div
      class="main"
    >
      <div
        class="text"
      >
        step.rolloutDeployment.timeout
      </div>
      <div
        class="bp3-popover-wrapper btnWrapper"
      >
        <span
          class="bp3-popover-target"
        >
          <button
            class="bp3-button bp3-minimal bp3-small StyledProps--font Button--button StyledProps--main copyIcon StyledProps--intent-primary"
            tabindex="0"
            type="button"
          >
            <span
              class="bp3-icon StyledProps--main"
              data-icon="copy"
            >
              <svg
                height="16"
                viewBox="0 0 14 16"
                width="16"
              >
                <g
                  fill="#22272D"
                  stroke="#22272D"
                  stroke-width="0.25"
                >
                  <path
                    d="M12.89.55H6.388a3.28 3.28 0 00-3.043 2.067C1.672 2.76.34 4.17.34 5.87v9.271c0 .17.138.308.307.308H7.15c1.35 0 2.543-.825 3.042-2.066 1.673-.144 3.006-1.554 3.006-3.255V.86a.307.307 0 00-.308-.31zM7.15 14.835H.955V5.87c0-1.299.955-2.4 2.199-2.624-.002.01-.001.021-.003.032-.031.182-.05.367-.05.552v9.27c0 .17.138.308.308.308h6.105a2.67 2.67 0 01-2.364 1.425zm5.432-4.705a2.68 2.68 0 01-2.607 2.664H3.716V3.831a2.672 2.672 0 012.671-2.665h6.195v8.963z"
                  />
                  <path
                    d="M6.11 6.672a.308.308 0 000 .615h1.732V9.02a.308.308 0 00.616 0V7.287h1.733a.308.308 0 000-.615H8.458V4.939a.308.308 0 00-.616 0v1.733H6.11z"
                  />
                </g>
              </svg>
            </span>
          </button>
        </span>
      </div>
    </div>
    <p
      class="StyledProps--font Text--lineclamp StyledProps--main"
      style="--text-line-clamp: 1;"
    >
      10m
    </p>
    <div
      class="main"
    >
      <div
        class="text"
      >
        step.rolloutDeployment.skipDryRun
      </div>
      <div
        class="bp3-popover-wrapper btnWrapper"
      >
        <span
          class="bp3-popover-target"
        >
          <button
            class="bp3-button bp3-minimal bp3-small StyledProps--font Button--button StyledProps--main copyIcon StyledProps--intent-primary"
            tabindex="0"
            type="button"
          >
            <span
              class="bp3-icon StyledProps--main"
              data-icon="copy"
            >
              <svg
                height="16"
                viewBox="0 0 14 16"
                width="16"
              >
                <g
                  fill="#22272D"
                  stroke="#22272D"
                  stroke-width="0.25"
                >
                  <path
                    d="M12.89.55H6.388a3.28 3.28 0 00-3.043 2.067C1.672 2.76.34 4.17.34 5.87v9.271c0 .17.138.308.307.308H7.15c1.35 0 2.543-.825 3.042-2.066 1.673-.144 3.006-1.554 3.006-3.255V.86a.307.307 0 00-.308-.31zM7.15 14.835H.955V5.87c0-1.299.955-2.4 2.199-2.624-.002.01-.001.021-.003.032-.031.182-.05.367-.05.552v9.27c0 .17.138.308.308.308h6.105a2.67 2.67 0 01-2.364 1.425zm5.432-4.705a2.68 2.68 0 01-2.607 2.664H3.716V3.831a2.672 2.672 0 012.671-2.665h6.195v8.963z"
                  />
                  <path
                    d="M6.11 6.672a.308.308 0 000 .615h1.732V9.02a.308.308 0 00.616 0V7.287h1.733a.308.308 0 000-.615H8.458V4.939a.308.308 0 00-.616 0v1.733H6.11z"
                  />
                </g>
              </svg>
            </span>
          </button>
        </span>
      </div>
    </div>
    <p
      class="StyledProps--font Text--lineclamp StyledProps--main"
      style="--text-line-clamp: 1;"
    />
  </div>
</div>
`;
